### ISSUE TITLE
Coveralls not working?

### ISSUE THREAD

**Issue by adnanhemani** [data-8/datascience#356]
It seems like Coveralls isn't actually tracking code coverage anymore....

It thinks the latest commit on master was back in 2016. Am I just reading this wrong or is this something we need to fix?

**Comment by davidwagner** [comment]
@SamLau95, do you know what's going on?  Do we need to poke coveralls somehow?

I signed into coveralls.io with my Github account, and I can view https://coveralls.io/github/data-8/datascience, but I can't see any option to get it to rerun coverage or to find the newer Travis builds.  I don't know if Sam or John have more access on coveralls than I do, though.

The most recent coveralls build is #924 on 9 Jul 2018 (https://coveralls.io/builds/17905611, https://travis-ci.org/data-8/datascience/jobs/402003675).   I can see that after Travis build #924, the next bunch of Travis builds all failed (if I recall correctly, it's because the underlying infrastructure updated to a newer version of Python which caused problems).  Only in the past 12 hours do we have passing Travis builds.  Perhaps coveralls ignored all the failed Travis builds, and then hasn't noticed the passing Travis builds yet?  Not sure how to wake up coveralls and get it to pay attention to the latest builds.

(I think it thinks the latest is 9 Jul 2018, not 2016; but that's still very old.)

**Comment by davidwagner** [comment]
@SamLau95 ,  Well, updating to the latest version of the libraries (commit 13817b) didn't seem to suffice.  We've had a few more Travis builds and nothing shows up on coveralls.  Thanks for giving that a try.

Do we need to set some configuration setting on Travis or on Github to set a webhook to notify coveralls.io for each new build / push to the repo? (If not, how does coveralls.io know when to check Travis for a new build?)

I don't think I have admin access on coveralls.io or Travis for the datascience project so I'm not sure how to look into this any further myself.

**Comment by SamLau95** [comment]
Sadly, I'm not sure why coveralls isn't updating. The way coveralls works is that Travis runs:

```
coverage run setup.py test
```

Which generates the coverage report. Then Travis runs the `coveralls` command which uploads the report to coveralls.

When coveralls last updated successfully, there was a confirmation message in Travis (https://travis-ci.org/data-8/datascience/builds/402003674#L2137-L2141):

```
$ coveralls
Submitting coverage to coveralls.io...
Coverage submitted!
Job #924.1
https://coveralls.io/jobs/38236748
```

However, every build after #924 gives this output (https://travis-ci.org/data-8/datascience/builds/543977273#L2401-L2403):

```
$ coveralls
Submitting coverage to coveralls.io...
29724
```

I'm not sure what to make of the `29724` output — could it be an error code?

As an aside, I also notice that the command after that to update the docs (`bash tools/deploy_docs.sh`) is breaking because of an authentication issue.

**Comment by davidwagner** [comment]
Thanks for explaining how this all works!  I ran coveralls locally and get the same `29724` output.  Puzzling.  I looked through all the coveralls docs I could find and couldn't find anything useful, so I filed an issue with the coveralls project: https://github.com/coveralls-clients/coveralls-python/issues/202

**Comment by davidwagner** [comment]
Incidentally, setting `ignore_errors = False` in the `.coveragerc` didn't change anything.

**Comment by davidwagner** [comment]
Thanks for mentioning the issue with `deploy_docs.sh`!  I filed a separate issue for that: #363

**Comment by davidwagner** [comment]
I suspect the issue is related to the version of the dependencies.

I reproduced this same `29724` output locally, on my own personal machine.  However I later discovered that my personal machine had been using very old versions of `coverage` and `coveralls` (namely, coverage 3.7.1 and coveralls 0.5).  When I wiped all old versions and installed the latest versions afresh (coverage 4.5.3 and coveralls 1.8.0), the problem goes away, and I can now see a coverage report successfully submitted to coveralls.io (namely, https://coveralls.io/jobs/49544400).  I did have to create a `.coveralls.yml` containing `repo_token: [..coveralls secret token here..]`, but then it worked locally.

So it appears that upgrading coverage and coveralls might suffice to fix the problem.  Now we just have to figure out why this problem still happens on Travis, despite commit 13817b01dd updating `setup.py` to specify the latest versions of coverage and coveralls.  Perhaps there's nothing in the Travis scripts to update the conda/pip packages for coveralls and coverage?  I notice that coverage and coveralls aren't mentioned in `linux_environment.yml`, so perhaps the `conda env update...` command isn't updating them -- maybe we need to add them to `linux_environment.yml`.  Let me see if I can figure out how to tickle Travis to tell me what version of coveralls and coverage are installed in the Travis image.

**Comment by davidwagner** [comment]
Yup, that was the problem (see https://travis-ci.org/data-8/datascience/builds/544494100).  Let me create a pull request for the fix.

**Comment by adnanhemani** [comment]
Reopening this issue - in Travis, we're seeing this error now. Have we seen this before?
<img width="1050" alt="Screen Shot 2019-06-15 at 3 03 12 PM" src="https://user-images.githubusercontent.com/10746770/59549978-b5c8d580-8f7e-11e9-875c-91854d89d1a6.png">

**Comment by davidwagner** [comment]
I think I remember seeing that when `coveralls` or `coverage` wasn't up-to-date, so maybe we should check the versions of those packages.

**Comment by adnanhemani** [comment]
Made a new PR #378 to fix this!
